File: src/14_Tree_Algorithms/binary_search_tree.cpp
#include <bits/stdc++.h>
using namespace std;


class Node
{
  public:
  int data;
  Node* left = nullptr;
  Node* right = nullptr;

  Node(int d): data(d) {};

  void insert(int value)
  {
    if (value <= data) {
      if (left) left-> insert(value);
      else left = new Node(value);
    } else {
      if (right) right-> insert(value);
      else right = new Node(value);
    }
  }

  bool contains(int value)
  {
    if (value == data) return true;
    if (value < data) {
      if (left) return left-> contains(value);
      else return false;
    } else {
      if (right) return right-> contains(value);
      else return false;
    }
  }

  void print_in_order()
  {
    if (left) left-> print_in_order();
    cout << data << ' ';
    if (right) right-> print_in_order();
  }

  void print_pre_order()
  {
    cout << data << ' ';
    if (left) left-> print_pre_order();
    if (right) right-> print_pre_order();
  }

  void print_post_order()
  {
    if (left) left-> print_post_order();
    if (right) right-> print_post_order();
    cout << data << ' ';
  }
};


Node* insert(Node* root, int data)
{
  if (!root) return new Node(data);
  if (data > root-> data) root-> right = insert(root-> right, data);
  else root-> left = insert(root-> left, data);
  return root;
}


const int n = 7;
int arr[n] = {5, 3, 4, 6, 2, 7, 1};
/*
 *          5
 *         / \
 *        3   6
 *       / \   \
 *      2   4   7
 *     /
 *    1
 */


int main()
{
  Node* root = new Node(arr[0]);
  for (int u = 1; u < n; ++u) root-> insert(arr[u]);
  // for (int u = 1; u < n; ++u) root = insert(root, arr[u]);

  cout << "In-order: ";
  root-> print_in_order();
  cout << '\n';
  cout << "Pre-order: ";
  root-> print_pre_order();
  cout << '\n';
  cout << "Post-order: ";
  root-> print_post_order();
  cout << '\n';
}

File: src/14_Tree_Algorithms/binary_search_tree.py
class _Node:

  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

  def insert(self, data):
    if data <= self.data:
      if self.left:
        self.left.insert(data)
      else:
        self.left = _Node(data)
    else:
      if self.right:
        self.right.insert(data)
      else:
        self.right = _Node(data)

  def contains(self, data):
    if self.data == data:
      return True
    if data <= self.data:
      if self.left:
        return self.left.contains(data)
      else:
        return False
    else:
      if self.right:
        return self.right.contains(data)
      else:
        return False


class BinaryTree:

  def __init__(self):
    self.root = None

  def insert(self, data):
    if self.root:
      self.root.insert(data)
    else:
      self.root = _Node(data)

  def contains(self, data):
    if self.root:
      return self.root.contains(data)
    return False

  def traverse_pre_order(self):
    yield from self._traverse_pre_order(self.root)

  def traverse_in_order(self):
    yield from self._traverse_in_order(self.root)

  def traverse_post_order(self):
    yield from self._traverse_post_order(self.root)

  @staticmethod
  def _traverse_pre_order(node):
    if node:
      yield node.data
      yield from BinaryTree._traverse_pre_order(node.left)
      yield from BinaryTree._traverse_pre_order(node.right)

  @staticmethod
  def _traverse_in_order(node):
    if node:
      yield from BinaryTree._traverse_in_order(node.left)
      yield node.data
      yield from BinaryTree._traverse_in_order(node.right)

  @staticmethod
  def _traverse_post_order(node):
    if node:
      yield from BinaryTree._traverse_post_order(node.left)
      yield from BinaryTree._traverse_post_order(node.right)
      yield node.data


if __name__ == '__main__':
  arr = [5, 3, 4, 6, 2, 7, 1]
  tree = BinaryTree()
  for x in arr:
    tree.insert(x)

  print("Pre-order :", ' '.join(map(str, tree.traverse_pre_order())))
  print("In-order  :", ' '.join(map(str, tree.traverse_in_order())))
  print("Post-order:", ' '.join(map(str, tree.traverse_post_order())))

  print()
  for x in [3, 4, 8]:
    print(f"tree.contains({x}) {tree.contains(x)}")
